### ISSUE TITLE
Disable spam/virus check for outbound messages

### ISSUE THREAD

**Issue by marcinkunert** [docker-mailserver/docker-mailserver#3170]
This is actually a follow-up for #1713. I've finally found some time to tinker on it. I would share my solution and ask a few questions about possible improvements.

I've enabled avamis and spamassasin in the env, so starting up both incoming and outgoing messages are checked for spam.

```
ENABLE_AMAVIS=1
ENABLE_SPAMASSASSIN=1
```

So basically every message goes through postfix. I don't quite understand why after receiving the message by `dovecot` it gets transferred to `postfix`, and not directly to the inbox. I guess it's about reusing some useful parts like queues, spamchecks etc.

The solution I've found is to route the traffic either to `amavis` or a dummy filter to send out the message.

What is important here is the `content_filter` in `/etc/postfix/main.cf`

```
content_filter = smtp-amavis:[127.0.0.1]:10024
```

So the traffic always goes to `smtp-amavis` for it to handle the check and send the message. We can override this by using a `FILTER` like this:

```
smtpd_sender_restrictions =
  check_sender_access FILTER smtp-amavis:[127.0.0.1]:10026
```

This would also use `smtp-amavis`, but on another port, which we could use to our advantage and disable some checks there, like this (config/amavis.cf):

```
$inet_socket_port = [10024, 10026];

#interface_policy{'10026'} = 'ORIGINATING';

$policy_bank{'ORIGINATING'} = {
    originating => 1, # declare that mail was submitted by our smtp client

    # don't perform spam/virus/header check.
    bypass_spam_checks_maps => [1],
    bypass_virus_checks_maps => [1],
    bypass_header_checks_maps => [1],

    # allow sending any file names and types
    bypass_banned_checks_maps => [1],

    final_spam_destiny   => D_PASS, # insure spam passes
    final_banned_destiny => D_PASS, # insure banned files pass
};
```

But this took ~400 ms per message, and I was not entirely happy with it. 

Ok, back to the topic. We need a conditional here. This can be used to switch between originating and foreign:

```
smtpd_sender_restrictions = 
  check_sender_access regexp:/etc/postfix/tag_as_originating.re
  permit_mynetworks
  permit_sasl_authenticated
  permit_tls_clientcerts
  reject_unknown_sender_domain
  reject_unknown_client_hostname
  check_sender_access regexp:/etc/postfix/tag_as_foreign.re
```

And the filters looks like this.
```
echo '/^/  FILTER filter:dummy ' > /etc/postfix/tag_as_originating.re
echo '/^/  FILTER smtp-amavis:[127.0.0.1]:10024' > /etc/postfix/tag_as_foreign.re
```

The missing part is the filter. It should be created in `master.cf`

```
filter    unix  -       n       n       -       10      pipe
    flags=Rq user=filter null_sender=
    argv=/opt/filter.sh -f ${sender} -- ${recipient}
```

It basically executes the `/opt/filter.sh` script:

```
#!/bin/sh

# Simple shell-based filter. It is meant to be invoked as follows:
#       /path/to/script -f sender recipients...

# Localize these. The -G option does nothing before Postfix 2.3.
INSPECT_DIR=/var/spool/filter
SENDMAIL="/usr/sbin/sendmail -G -i" # NEVER NEVER NEVER use "-t" here.

# Exit codes from <sysexits.h>
EX_TEMPFAIL=75
EX_UNAVAILABLE=69

# Clean up when done or when aborting.
trap "rm -f in.$$" 0 1 2 3 15

# Start processing.
cd $INSPECT_DIR || {
    echo $INSPECT_DIR does not exist; exit $EX_TEMPFAIL; }

cat >in.$$ || {
    echo Cannot save mail to file; exit $EX_TEMPFAIL; }

$SENDMAIL "$@" <in.$$

exit $?
```

Finally, to the questions:

- the filter seems like a strange workaround, couldn't an existing service declared in `/etc/postfix/master.cf` used instead? I've tried using `smtpd:10025`, but it would fail from time to time with "corrupted format" message (or something like that)
- the filter sends it using `sendmail`, does postfix also use it? I'm not quite sure about the relationship between these two? I assume sendmail is a part of postfix`
- in the ideal case I would love to remove the entire filter part, I wonder if the `content_filter = smtp-amavis:[127.0.0.1]:10024` can be cancelled out completely - I have to check that
- I can't understand how the `smtpd_sender_restrictions` works. I guess it looks for the first `true` value, so after finding `permit_mynetworks` to be `true` it doesn't process the params further
- do you see any points of what may fail here?

**Comment by polarathene** [comment]
> What is important here is the `content_filter` in `/etc/postfix/main.cf`
>I wonder if the `content_filter = smtp-amavis:[127.0.0.1]:10024` can be cancelled out completely - I have to check that

`master.cf` defines separate services for ports 587 (`submission`) and 465 (`smtps`, technically should be `submissions`), these could add a line each to opt-out of the filtering as they're intended for submitting mail to deliver outbound (or locally), while port 25 is meant for receiving inbound mail from external MTAs:

https://github.com/docker-mailserver/docker-mailserver/blob/0baf82f7d0d1f160edef86012ebcac4131ea9ebe/target/postfix/master.cf#L44

---

Dovecot does not receive mail sent to the mail-server, it's Postfix job to handle all that, and pass over local mail to Dovecot via LMTP. Dovecot provides access to retrieve / interact with that mail storage via IMAP and POP3.

Postfix would receive mail, and if not rejected pass it over to Amavis to check on an inbound port 10024, and return it back to Postfix on port 10025 IIRC, so that it's not going through a loop. 

These are [defined in `master.cf` just like the submission(s) services](https://github.com/docker-mailserver/docker-mailserver/blob/0baf82f7d0d1f160edef86012ebcac4131ea9ebe/target/amavis/postfix-amavis.cf) mentioned earlier, but maintained in a separate file where we merge it in during startup (since v12, presently unreleased). `smtp-amavis` was defined in `main.cf:content_filter` to reference the equivalent `smtp-amavis` service in `main.cf` that I linked to.

I'm not sure why an IP and port is defined in `main.cf:content_filter` while the `master.cf` seems to indicate it as a [`unix` socket instead of `inet` TCP socket type](https://www.postfix.org/master.5.html) (those docs describe the `inet` format too). So that's potentially misconfigured but still working AFAIK :man_shrugging: (_`127.0.0.1:10024` depends on the `amavis` service to be running, but is still [relevant to Postfix `master.cf`](https://wiki.archlinux.org/title/Amavis#Quick_start) apparently_)

---

> I can't understand how the `smtpd_sender_restrictions` works. I guess it looks for the first `true` value, so after finding `permit_mynetworks` to be `true` it doesn't process the params further

As far as I understand it, yes that sounds right. IIRC, Postfix does document configuring restrictions for spam handling separate from relay restrictions which might be relevant to you.

AFAIK though, `content_filter` is only going to apply to `smtpd` (Postfix receiving mail) not `smtp` (Postfix sending mail), so using the `content_filter=` to opt-out like shown at the start of this message on the submission(s) services and sending mail through those ports, they should not go through any spam filtering.

**Comment by casperklein** [comment]
I can confirm, outgoing mails submitted via port 465 are flagged as spam. Tested with the [GTUBE](https://en.wikipedia.org/wiki/GTUBE).

I've added `-o content_filter=` above this line https://github.com/docker-mailserver/docker-mailserver/blob/0baf82f7d0d1f160edef86012ebcac4131ea9ebe/target/postfix/master.cf#L18

and restarted postfix. The messages are still flagged as spam.

**Comment by polarathene** [comment]
> I can confirm, outgoing mails submitted via port 465 are flagged as spam. Tested with the [GTUBE](https://en.wikipedia.org/wiki/GTUBE).
> 
> I've added `-o content_filter=` above this line

You verified on port 465, but modified port 587? `smtps` is port 465.

I've been meaning to update `master.cf` there to replace `smtps` with `submissions`, they're both for port 465 and mapped in `/etc/services`, with `smtps` being the legacy mapping prior to 2018. No functional difference though.

**Comment by casperklein** [comment]
> You verified on port 465, but modified port 587? smtps is port 465.

Thanks. I overlooked that. However, even setting `content_filter` in the right section, outgoing mails are still marked as spam.

https://github.com/docker-mailserver/docker-mailserver/blob/0baf82f7d0d1f160edef86012ebcac4131ea9ebe/target/postfix/master.cf#L31

> I've been meaning to update master.cf there to replace smtps with submissions

This way, my error could have been avoided 😆 On the other hand, `smtps` keeps our config "in sync" with the default config (on debian) which uses `smtps`.

**Comment by polarathene** [comment]
> However, even setting `content_filter` in the right section, outgoing mails are still marked as spam.

I don't have the bandwidth to look into it further personally. I assumed `content_filter` is what pass it through Amavis, perhaps the `smtp` service would also need an opt-out if this setting is for both inbound (smtpd) and outbound (smtp) :man_shrugging: 

I'm also not sure if it's worth it if @georglauterbach plans to drop Amavis in favour of Rspamd before the end of this year.

> keeps our config "in sync" with the default config (on debian)

We already stray from that a fair bit now with the opt-out of chroot which matches official Postfix config more. Plus formatting changes. Not sure how useful the diff would be between the two now, but I doubt `smtps` vs `submissions` would add much more noise :man_shrugging: 

Personally, I rather be more distro agnostic and [align better with the projects config decisions](https://github.com/vdukhovni/postfix/blob/3474213a2b033916ae622ae985768031aad4e2b0/postfix/conf/master.cf#L38), which as you can see use `submissions`. Last use of `smtps` in that file was April 2021 (_[replaced with `submissions` in May 2021 for Postfix 3.7](https://github.com/vdukhovni/postfix/commit/c0e76bc1f16273f24a4820020f0958d62804a4f2#diff-9e014842e28d87358be8ddb3a2b7531b6f96af5e81c59fbf9af986e773ac329cR25563-R25566)_), so Debian will eventually catch-up I imagine.

**Comment by georglauterbach** [comment]
This is implemented now in v13.0.0 but for Rspamd only. I doubt we will integrate it for Amavis/SA too (maintainers simply lack time to do so). I will close this as you can use Rspamd and this feature.
